src/circular_queue.cpp

```diff
@@ -0,0 +1,78 @@
+#include "../include/circular_queue.h"
+#include "../include/update.h"
+#include "../include/buffer_tree.h"
+
+#include <string.h>
+#include <chrono>
+
+CircularQueue::CircularQueue(int num_elements, int size_of_elm): 
+  len(num_elements), elm_size(size_of_elm) {
+	head = 0;
+	tail = 0;
+
+	// malloc the memory for the circular queue
+	queue_array = (queue_elm *) malloc(sizeof(queue_elm) * len);
+	for (int i = 0; i < len; i++) {
+		queue_array[i].data  = (char *) malloc(elm_size * sizeof(char));
```

Comment:
did you consider alternatively housing the data buffer in the same structure as your queue? my gut tells me this might be more efficient because of data locality

Reply:
Hmm good idea. So basically have one single data buffer which the queue_elm's point to?

Reply:
That's how I interpreted this. See most recent commit

Reply:
That's the idea. My hunch is this would save a tiny bit of time with cache pre-fetching but this may not be worth the extra implementation.

Reply:
It was an easy change actually

Reply:
and a good idea